<h1>Содержание<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Подготовка" data-toc-modified-id="Подготовка-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Подготовка</a></span></li><li><span><a href="#Обучение" data-toc-modified-id="Обучение-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Обучение</a></span><ul class="toc-item"><li><span><a href="#Модель-логической-регрессии" data-toc-modified-id="Модель-логической-регрессии-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>Модель логической регрессии</a></span></li><li><span><a href="#Модель-решающего-древа" data-toc-modified-id="Модель-решающего-древа-2.2"><span class="toc-item-num">2.2&nbsp;&nbsp;</span>Модель решающего древа</a></span></li><li><span><a href="#Модель-CatBoostClassifier" data-toc-modified-id="Модель-CatBoostClassifier-2.3"><span class="toc-item-num">2.3&nbsp;&nbsp;</span>Модель CatBoostClassifier</a></span></li></ul></li><li><span><a href="#Выводы" data-toc-modified-id="Выводы-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Выводы</a></span></li><li><span><a href="#Чек-лист-проверки" data-toc-modified-id="Чек-лист-проверки-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Чек-лист проверки</a></span></li></ul></div>

# Проект для «Викишоп»

Интернет-магазин запускает новый сервис. Теперь пользователи могут редактировать и дополнять описания товаров, как в вики-сообществах. То есть клиенты предлагают свои правки и комментируют изменения других. Магазину нужен инструмент, который будет искать токсичные комментарии и отправлять их на модерацию. 

Обучите модель классифицировать комментарии на позитивные и негативные. В вашем распоряжении набор данных с разметкой о токсичности правок.

## Подготовка

### Имортируем все необходимые инструмены

In [1]:
import matplotlib.pyplot as plt
import numpy as np
import pandas as pd
import re
import swifter
import spacy
from sklearn.pipeline import Pipeline
from tqdm.notebook import tqdm
tqdm.pandas()
from pymystem3 import Mystem
from nltk.tokenize import word_tokenize
import nltk
from nltk.corpus import stopwords as nltk_stopwords
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.model_selection import (
    train_test_split,
    GridSearchCV
)
from sklearn.linear_model import LogisticRegression
from sklearn.tree import DecisionTreeClassifier
from sklearn.feature_extraction.text import CountVectorizer
from sklearn.feature_extraction.text import TfidfTransformer
from nltk.stem import WordNetLemmatizer
from nltk.corpus import wordnet
from catboost import CatBoostClassifier
from sklearn.metrics import f1_score
import warnings
warnings.filterwarnings('ignore')

### Загрузим данные

In [2]:
try:
    data = pd.read_csv("toxic_comments.csv", index_col=0)
except:
    data = pd.read_csv('D:\\toxic_comments.csv', index_col=0)

### Откроем первые 5 строк каждой таблицы

In [3]:
data.sample(5)

text  toxic
158117  No, it's not a delayed reaction\n\nI just happ...      0
56222   "\n\n Category:Atlantic Records  \n\nHow the h...      0
78947   Belarus flag \n\nSorry aout that. I was in the...      0
86069   Hi there from PORTUGAL,\n\nPlease leave just A...      0
20587   "\n\nOH, OH, NICE ONE, REVERTING MY EDITS! CHR...      1

### Посмотрим информацию о данных

In [4]:
data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 159292 entries, 0 to 159450
Data columns (total 2 columns):
 #   Column  Non-Null Count   Dtype 
---  ------  --------------   ----- 
 0   text    159292 non-null  object
 1   toxic   159292 non-null  int64 
dtypes: int64(1), object(1)
memory usage: 3.6+ MB


### Посмотрим баланс у целевого признака

In [5]:
data['toxic'].value_counts()

0    143106
1     16186
Name: toxic, dtype: int64

**Наблюдаем дисбаланс**

### Приведём текст к нижнему регистру

In [6]:
data['text'] = data['text'].str.lower()

### Лемматизируем текст в колонке с признаком

In [7]:
nlp = spacy.load("en_core_web_sm")

In [8]:
def clear_text(text):
    lists = re.sub(r'[^a-zA-Z ]', ' ', text.lower())
    words = lists.split()
    words = " ".join(words) 
    return words

In [9]:
def lemma_clear(text): 
      
    lemm = nlp(text) 
    lemm = " ".join([token.lemma_ for token in lemm])  
    return " ".join(lemm.split())

In [10]:
data['lemm_text'] = data['text'].apply(clear_text)

In [11]:
data['lemm_text'] = data['lemm_text'].apply(lemma_clear)

In [12]:
data.head()

text  toxic  \
0  explanation\nwhy the edits made under my usern...      0   
1  d'aww! he matches this background colour i'm s...      0   
2  hey man, i'm really not trying to edit war. it...      0   
3  "\nmore\ni can't make any real suggestions on ...      0   
4  you, sir, are my hero. any chance you remember...      0   

                                           lemm_text  
0  explanation why the edit make under my usernam...  
1  d aww he match this background colour I m seem...  
2  hey man I m really not try to edit war it s ju...  
3  more I can t make any real suggestion on impro...  
4  you sir be my hero any chance you remember wha...

### Подготовим данные к обучению моделей

In [13]:
target = data['toxic']
features = data['lemm_text']

In [14]:
features_train, features_test, target_train, test_target = train_test_split(
    features, target, test_size=0.4, random_state=12345, stratify=target)

features_valid, features_test, target_valid, test_target = train_test_split(
    features_test, test_target, test_size=0.50, random_state=12345, stratify=test_target)

features_train.shape, features_valid.shape, features_test.shape

((95575,), (31858,), (31859,))

### Уберём в признаках стоп-слова

In [15]:
nltk.download('stopwords')
stopwords = set(nltk_stopwords.words('english'))

[nltk_data] Downloading package stopwords to
[nltk_data]     C:\Users\piros\AppData\Roaming\nltk_data...
[nltk_data]   Package stopwords is already up-to-date!


In [16]:
count_tf_idf = TfidfVectorizer(stop_words=list(stopwords))

In [17]:
tf_idf_train = count_tf_idf.fit_transform(features_train)
tf_idf_test = count_tf_idf.transform(features_test)
tf_idf_valid = count_tf_idf.transform(features_valid)

## Обучение

### Модель логической регрессии

In [18]:
scoring = 'f1'
CV = 4

pipeline_logr = Pipeline([
    ('vectorizer', TfidfVectorizer()),
    ('classifier', LogisticRegression(random_state=12345))
])

params = {
    'vectorizer__ngram_range': [(1,1), (1,2)],
    'classifier__penalty': ['l1', 'l2'],
    'classifier__C': [0.01, 0.1, 1, 10, 100]
}

grid = GridSearchCV(pipeline_logr, cv=CV, n_jobs=-1, param_grid=params ,scoring=scoring)
grid.fit(features_train, target_train)
grid.score(features_test, test_target)

0.7781793150920667

**Модель логической регрессии показала лучший результат после использование твоих советов, спасибо:**

***На тестовой выборке 0.77***

### Модель решающего древа

In [19]:
model_tree = DecisionTreeClassifier(
            random_state=12345,
            max_depth=8,
            min_samples_split=10,
            min_samples_leaf=10,
            class_weight='balanced'
            )
model_tree.fit(tf_idf_train, target_train)
predictions = model_tree.predict(tf_idf_test)
f1_score(test_target, predictions)

0.5254382214621633

**К сожалению решающее древо показало плохой результат на тестовой выборке 0.52**

### Модель CatBoostClassifier

In [20]:
best_result = 0
best_depth = 0
best_iterations = 0
for depth in range(2, 10, 2):
    for iterations_i in range(100, 200, 100):
        model_cat = CatBoostClassifier(
                                      random_state=12345,
                                      task_type='CPU',
                                      iterations=iterations_i,
                                      depth = depth,
                                      od_type='Iter'
                                      )
        model_cat.fit(tf_idf_train, target_train)
        predictions = model_cat.predict(tf_idf_test)
        f1 = f1_score(test_target, predictions)
        if best_result < f1:
            best_result = f1
            best_depth = depth
            best_iterations = iterations_i
print(best_result)
print(best_depth)
print(best_iterations)

Learning rate set to 0.5
0:	learn: 0.3561758	total: 260ms	remaining: 25.7s
1:	learn: 0.2811307	total: 352ms	remaining: 17.2s
2:	learn: 0.2583851	total: 441ms	remaining: 14.3s
3:	learn: 0.2467951	total: 527ms	remaining: 12.6s
4:	learn: 0.2388936	total: 620ms	remaining: 11.8s
5:	learn: 0.2331570	total: 706ms	remaining: 11.1s
6:	learn: 0.2290555	total: 798ms	remaining: 10.6s
7:	learn: 0.2258873	total: 887ms	remaining: 10.2s
8:	learn: 0.2198779	total: 978ms	remaining: 9.89s
9:	learn: 0.2165926	total: 1.06s	remaining: 9.59s
10:	learn: 0.2133789	total: 1.16s	remaining: 9.36s
11:	learn: 0.2104942	total: 1.24s	remaining: 9.12s
12:	learn: 0.2077081	total: 1.33s	remaining: 8.91s
13:	learn: 0.2043322	total: 1.42s	remaining: 8.72s
14:	learn: 0.2025165	total: 1.51s	remaining: 8.55s
15:	learn: 0.2005247	total: 1.6s	remaining: 8.38s
16:	learn: 0.1986823	total: 1.69s	remaining: 8.24s
17:	learn: 0.1953375	total: 1.78s	remaining: 8.12s
18:	learn: 0.1938129	total: 1.87s	remaining: 7.98s
19:	learn: 0.1919

In [21]:
model = CatBoostClassifier(
    random_state=12345,
    task_type='CPU',
    iterations=100,
    depth = 8,
    od_type='Iter'                                   
)
predictions = model_cat.predict(tf_idf_valid)
f1_score(target_valid, predictions)


0.7621908127208481

**CatBoostRegressor с показателями:**

***На тестовой выборке F1 = 0.74***

***На валидационной выборке F1 = 0.76***

## Выводы

**Были проведенные исследования данных и были выполненны следующие действия:**

**Данные были загруженны**

**Так же данные были лемматизированны**

**Данные были подготовленны к обучению моделий, а так же из них были удалены стоп-слова**

**Были обучены три модели и они показали следующий результат:**

***Модель логической регрессии - F1 = 0.7791***

***Модель решающего древа - F1 = 0.52***

***Модель CatBoostClassifier:***

***На тестовой выборке F1 = 0.74***

***На валидационной выборке F1 = 0.76***

**Модель логической регрессии показала себя лучшей среди тестируемых c F1 = 0.77**